Libplanet/Action/ActionEvaluator.cs

```diff
@@ -0,0 +1,581 @@
+            StateCompleterSet<T> stateCompleterSet)
+        {
+            ITrie? previousBlockStatesTrie = !(_trieGetter is null) && block.PreviousHash is { } h
+                ? _trieGetter(h)
+                : null;
+            IAccountStateDelta previousStates =
+                GetPreviousBlockOutputStates(block, stateCompleterSet);
+
+            ImmutableList<ActionEvaluation> evaluations = EvaluateBlock(
+                block: block,
+                currentTime: DateTimeOffset.UtcNow,
+                previousStates: previousStates,
+                previousBlockStatesTrie: previousBlockStatesTrie).ToImmutableList();
+
+            if (_policyBlockAction is null)
+            {
+                return evaluations;
+            }
+            else
+            {
+                previousStates = evaluations.Count > 0
+                    ? evaluations.Last().OutputStates
+                    : previousStates;
+                return evaluations.Add(
+                    EvaluatePolicyBlockAction(
+                        block: block,
+                        previousStates: previousStates,
+                        previousBlockStatesTrie: previousBlockStatesTrie));
+            }
+        }
+
+        /// <summary>
+        /// Retrieves the set of <see cref="Address"/>es that will be updated when
+        /// a given <see cref="Transaction{T}"/> is evaluated.
+        /// </summary>
+        /// <param name="tx">The <see cref="Transaction{T}"/> to evaluate.</param>
+        /// <returns>An <see cref="IImmutableSet{T}"/> of updated <see cref="Address"/>es.
+        /// </returns>
+        /// <remarks>
+        /// A mock evaluation is performed on <paramref name="tx"/> using a mock
+        /// <see cref="Block{T}"/> for its evaluation context and a mock
+        /// <see cref="IAccountStateDelta"/> as its previous state to obtain the
+        /// <see cref="IImmutableSet{T}"/> of updated <see cref="Address"/>es.
+        /// </remarks>
+        internal static IImmutableSet<Address> GetUpdatedAddresses(Transaction<T> tx)
+        {
+            IAccountStateDelta previousStates = new AccountStateDeltaImpl(
+                _nullAccountStateGetter,
+                _nullAccountBalanceGetter,
+                tx.Signer);
+            IEnumerable<ActionEvaluation> evaluations = ActionEvaluator<T>.EvaluateActions(
+                preEvaluationHash: _nullBlock.PreEvaluationHash,
+                blockIndex: _nullBlock.Index,
+                txid: tx.Id,
+                previousStates: previousStates,
+                miner: _nullBlock.Miner.GetValueOrDefault(),
+                signer: tx.Signer,
+                signature: tx.Signature,
+                actions: tx.Actions.Cast<IAction>().ToImmutableList(),
+                rehearsal: true,
+                previousBlockStatesTrie: null);
+
+            if (evaluations.Any())
+            {
+                return evaluations.Last().OutputStates.UpdatedAddresses;
+            }
```

Comment:
Trivial
```suggestion
            if (evaluations.LastOrDefault() is { } lastEval)
            {
                return lastEval.OutputStates.UpdatedAddresses;
            }
```

Reply:
Hmm... haven't really put my thought into it when writing this, but looking back, wouldn't this be problematic if for whatever reason `default` is implemented and `AccountStateDeltaImpl` becomes a `struct`? I know it is clear from the code that `AccountStateDeltaImpl` is a `class` (and also it implements an `interface`, which also makes it "likely" to be a `class`) but conceptually, I don't think an evaluation output should necessarily be a `class`.

Reply:
Fair enough.